Error building latest git on FreeBSD

0. Reviewed guide and contributing documents? (Yes/No): Yes
1. version [x.y.z, hash, other]: latest git clone
2. installed as a package or compiled from sources [deb, rpm, git, other]: source
3. standalone or part of third party [motion, MotionEyeOS, other]: standalone
4. video stream source [V4L (card or USB), net cam (mjpeg, rtsp, other), mmal]: net cam
5. hardware [x86, ARM, other]: x86
6. operating system [16.04, Stretch, etc, FreeBSD, other]: FreeBSD 12.1-RELEASE-p10 amd64

After a git pull to update my local sources to the latest revision I get the following errors during compile. My previous successful build was from a couple of months ago, and at that time it built w/o any problem.

$ git clone https://github.com/Motion-Project/motion.git
...
$ cd motion
$ autoreconf -fiv
...
$ ./configure
   **************************
      Configure status       
      motion 4.3.1+dirty20200928-21b6740
   **************************

CFLAGS: -D_THREAD_SAFE -I/usr/local/include  -I/usr/local/include  -I/usr/local/include  -I/usr/local/include  -g -O2

LIBS:  -pthread   -L/usr/local/lib -ljpeg  -L/usr/local/lib -lmicrohttpd  -L/usr/local/lib -lavutil -lavformat -lavcodec -lswscale -lavdevice  -L/usr/local/lib -lsqlite3 

LDFLAGS:  


OS                  : freebsd12.1
pthread_np          : yes
pthread_setname_np  : no
pthread_getname_np  : no
XSI error           : yes
webp support        : yes
V4L2 support        : yes
BKTR support        : yes
MMAL support        : no
FFmpeg support      : yes
libavformat version : 58.45.100
SQLite3 support     : yes
MYSQL support       : no
PostgreSQL support  : no
MariaDB support     : no

Install prefix:       /usr/local

$ gmake
...
cc -DHAVE_CONFIG_H -I. -I..  -Dsysconfdir=\"/usr/local/etc/motion\" -DLOCALEDIR=\"/usr/local/share/locale\" -I/usr/local/include -D_THREAD_SAFE -I/usr/local/include  -I/usr/local/include  -I/usr/local/include  -I/usr/local/include  -g -O2 -MT video_bktr.o -MD -MP -MF $depbase.Tpo -c -o video_bktr.o video_bktr.c &&\
mv -f $depbase.Tpo $depbase.Po
video_bktr.c:349:21: error: no member named 'update_parms' in 'struct params_context'; did you mean 'update_params'?
    if (!cnt->vdev->update_parms) return;
                    ^~~~~~~~~~~~
                    update_params
./motion.h:227:9: note: 'update_params' declared here
    int update_params;                        /*Bool for whether to update the parameters on the device*/
        ^
video_bktr.c:354:44: error: no member named 'usrctrl_count' in 'struct params_context'
    for (indx_user=0; indx_user<cnt->vdev->usrctrl_count; indx_user++){
                                ~~~~~~~~~  ^
video_bktr.c:355:29: error: no member named 'usrctrl_array' in 'struct params_context'
        usritem=&cnt->vdev->usrctrl_array[indx_user];
                 ~~~~~~~~~  ^
video_bktr.c:356:32: error: incomplete definition of type 'struct vdev_usrctrl_ctx'
        if (!strcasecmp(usritem->ctrl_name,"contrast")) bktr_set_contrast(dev,usritem->ctrl_value);
                        ~~~~~~~^
video_bktr.c:347:12: note: forward declaration of 'struct vdev_usrctrl_ctx'
    struct vdev_usrctrl_ctx *usritem;
           ^
video_bktr.c:356:86: error: incomplete definition of type 'struct vdev_usrctrl_ctx'
        if (!strcasecmp(usritem->ctrl_name,"contrast")) bktr_set_contrast(dev,usritem->ctrl_value);
                                                                              ~~~~~~~^
video_bktr.c:347:12: note: forward declaration of 'struct vdev_usrctrl_ctx'
    struct vdev_usrctrl_ctx *usritem;
           ^
video_bktr.c:357:32: error: incomplete definition of type 'struct vdev_usrctrl_ctx'
        if (!strcasecmp(usritem->ctrl_name,"hue")) bktr_set_hue(dev,usritem->ctrl_value);
                        ~~~~~~~^
video_bktr.c:347:12: note: forward declaration of 'struct vdev_usrctrl_ctx'
    struct vdev_usrctrl_ctx *usritem;
           ^
video_bktr.c:357:76: error: incomplete definition of type 'struct vdev_usrctrl_ctx'
        if (!strcasecmp(usritem->ctrl_name,"hue")) bktr_set_hue(dev,usritem->ctrl_value);
                                                                    ~~~~~~~^
video_bktr.c:347:12: note: forward declaration of 'struct vdev_usrctrl_ctx'
    struct vdev_usrctrl_ctx *usritem;
           ^
video_bktr.c:358:32: error: incomplete definition of type 'struct vdev_usrctrl_ctx'
        if (!strcasecmp(usritem->ctrl_name,"brightness")) bktr_set_brightness(dev,usritem->ctrl_value);
                        ~~~~~~~^
video_bktr.c:347:12: note: forward declaration of 'struct vdev_usrctrl_ctx'
    struct vdev_usrctrl_ctx *usritem;
           ^
video_bktr.c:358:90: error: incomplete definition of type 'struct vdev_usrctrl_ctx'
        if (!strcasecmp(usritem->ctrl_name,"brightness")) bktr_set_brightness(dev,usritem->ctrl_value);
                                                                                  ~~~~~~~^
video_bktr.c:347:12: note: forward declaration of 'struct vdev_usrctrl_ctx'
    struct vdev_usrctrl_ctx *usritem;
           ^
video_bktr.c:359:32: error: incomplete definition of type 'struct vdev_usrctrl_ctx'
        if (!strcasecmp(usritem->ctrl_name,"saturation")) bktr_set_saturation(dev,usritem->ctrl_value);
                        ~~~~~~~^
video_bktr.c:347:12: note: forward declaration of 'struct vdev_usrctrl_ctx'
    struct vdev_usrctrl_ctx *usritem;
           ^
video_bktr.c:359:90: error: incomplete definition of type 'struct vdev_usrctrl_ctx'
        if (!strcasecmp(usritem->ctrl_name,"saturation")) bktr_set_saturation(dev,usritem->ctrl_value);
                                                                                  ~~~~~~~^
video_bktr.c:347:12: note: forward declaration of 'struct vdev_usrctrl_ctx'
    struct vdev_usrctrl_ctx *usritem;
           ^
video_bktr.c:362:16: error: no member named 'update_parms' in 'struct params_context'; did you mean 'update_params'?
    cnt->vdev->update_parms = FALSE;
               ^~~~~~~~~~~~
               update_params
./motion.h:227:9: note: 'update_params' declared here
    int update_params;                        /*Bool for whether to update the parameters on the device*/
        ^
video_bktr.c:709:28: error: no member named 'usrctrl_count' in 'struct params_context'
            if (cnt->vdev->usrctrl_count > 0){
                ~~~~~~~~~  ^
video_bktr.c:710:45: error: no member named 'usrctrl_count' in 'struct params_context'
                for (indx=0;indx<cnt->vdev->usrctrl_count;indx++){
                                 ~~~~~~~~~  ^
video_bktr.c:711:37: error: no member named 'usrctrl_array' in 'struct params_context'
                    free(cnt->vdev->usrctrl_array[indx].ctrl_name);
                         ~~~~~~~~~  ^
video_bktr.c:712:32: error: no member named 'usrctrl_array' in 'struct params_context'
                    cnt->vdev->usrctrl_array[indx].ctrl_name=NULL;
                    ~~~~~~~~~  ^
video_bktr.c:715:24: error: no member named 'usrctrl_count' in 'struct params_context'
            cnt->vdev->usrctrl_count = 0;
            ~~~~~~~~~  ^
video_bktr.c:716:28: error: no member named 'usrctrl_array' in 'struct params_context'
            if (cnt->vdev->usrctrl_array != NULL){
                ~~~~~~~~~  ^
video_bktr.c:717:33: error: no member named 'usrctrl_array' in 'struct params_context'
                free(cnt->vdev->usrctrl_array);
                     ~~~~~~~~~  ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]

--- a/src/conf.c
+++ b/src/conf.c
@@ -2798,7 +2798,10 @@ static struct context **copy_video_params(struct context **cnt, const char *conf
         return cnt;
     }
 
-    if (strcmp(dep_config_params[config_indx].name,"power_line_frequency") &&
+    if ((!strcmp(dep_config_params[config_indx].name,"brightness") ||
+         !strcmp(dep_config_params[config_indx].name,"contrast") ||
+         !strcmp(dep_config_params[config_indx].name,"saturation") ||
+         !strcmp(dep_config_params[config_indx].name,"hue")) &&
         (parmval == 0)) {
         return cnt;
     }

--- a/src/motion.h
+++ b/src/motion.h
@@ -402,6 +402,10 @@ struct context {
     int diffs_last[THRESHOLD_TUNE_LENGTH];
     int smartmask_speed;
 
+    /* bktr and v4l2 params used for round robin */
+    int param_input;
+    long param_freq;
+    int param_norm;
 
     /* Commands to the motion thread */
     volatile unsigned int snapshot;    /* Make a snapshot */

--- a/src/video_bktr.c
+++ b/src/video_bktr.c
@@ -344,40 +344,37 @@ static void bktr_picture_controls(struct context *cnt, struct video_dev *viddev)
 {
 
     int dev = viddev->fd_device;
-    int indx_user, retcd;
-    struct vdev_usrctrl_ctx *usritem;
+    int indx;
+    struct params_item_ctx *usritem;
 
-    if (!cnt->vdev->update_parms) {
+    if (!cnt->vdev->update_params) {
         return;
     }
 
-    retcd = vid_parms_parse(cnt);
-    if (retcd < 0) {
-        return;
-    }
+    vid_parms_parse(cnt);
 
-    for (indx_user=0; indx_user<cnt->vdev->usrctrl_count; indx_user++) {
-        usritem=&cnt->vdev->usrctrl_array[indx_user];
-        if (!strcasecmp(usritem->ctrl_name,"contrast")) {
-            bktr_set_contrast(dev,usritem->ctrl_value);
+    for (indx = 0;indx < cnt->vdev->params_count; indx++) {
+        usritem=&cnt->vdev->params_array[indx];
+        if (!strcasecmp(usritem->param_name,"contrast")) {
+            bktr_set_contrast(dev,atoi(usritem->param_value));
         }
-        if (!strcasecmp(usritem->ctrl_name,"hue")) {
-            bktr_set_hue(dev,usritem->ctrl_value);
+        if (!strcasecmp(usritem->param_name,"hue")) {
+            bktr_set_hue(dev,atoi(usritem->param_value));
         }
-        if (!strcasecmp(usritem->ctrl_name,"brightness")) {
-            bktr_set_brightness(dev,usritem->ctrl_value);
+        if (!strcasecmp(usritem->param_name,"brightness")) {
+            bktr_set_brightness(dev,atoi(usritem->param_value));
         }
-        if (!strcasecmp(usritem->ctrl_name,"saturation")) {
-            bktr_set_saturation(dev,usritem->ctrl_value);
+        if (!strcasecmp(usritem->param_name,"saturation")) {
+            bktr_set_saturation(dev,atoi(usritem->param_value));
         }
     }
 
-    cnt->vdev->update_parms = FALSE;
+    cnt->vdev->update_params = FALSE;
 
 }
 
 static unsigned char *bktr_device_init(struct video_dev *viddev, int width, int height,
-                                unsigned input, unsigned norm, long freq)
+                                int input, int norm, long freq)
 {
     int dev_bktr = viddev->fd_device;
     struct sigaction act, old;
@@ -708,8 +705,7 @@ void bktr_cleanup(struct context *cnt)
 
         struct video_dev *dev = viddevs;
         struct video_dev *prev = NULL;
-        int indx;
-
+        
         /* Cleanup the v4l part */
         pthread_mutex_lock(&bktr_mutex);
 
@@ -727,20 +723,12 @@ void bktr_cleanup(struct context *cnt)
         cnt->video_dev = -1;
 
         /* free the information we collected regarding the controls */
+         /* free the information we collected regarding the controls */
         if (cnt->vdev != NULL) {
-            if (cnt->vdev->usrctrl_count > 0) {
-                for (indx=0;indx<cnt->vdev->usrctrl_count;indx++) {
-                    free(cnt->vdev->usrctrl_array[indx].ctrl_name);
-                    cnt->vdev->usrctrl_array[indx].ctrl_name=NULL;
-                }
-            }
-            cnt->vdev->usrctrl_count = 0;
-            if (cnt->vdev->usrctrl_array != NULL) {
-                free(cnt->vdev->usrctrl_array);
-                cnt->vdev->usrctrl_array = NULL;
-            }
+            util_parms_free(cnt->vdev);
 
             free(cnt->vdev);
+            cnt->vdev = NULL;
         }
 
         if (dev == NULL) {
@@ -815,8 +803,6 @@ int bktr_start(struct context *cnt)
         struct video_dev *dev;
         int bktr_fdtuner = -1;
         int width, height, bktr_method;
-        int input, norm;
-        long frequency;
         int fd_device = -1;
 
         MOTION_LOG(NTC, TYPE_VIDEO, NO_ERRNO, "[%s]", conf->video_device);
@@ -841,19 +827,6 @@ int bktr_start(struct context *cnt)
 
         width = conf->width;
         height = conf->height;
-
-        for (indx = 0; indx < cnt->vdev->params_count; indx++) {
-            if ( !strcmp(cnt->vdev->params_array[indx].param_name, "input")) {
-                input = atoi(cnt->vdev->params_array[indx].param_value);
-            }
-            if ( !strcmp(cnt->vdev->params_array[indx].param_name, "norm")) {
-                norm = atoi(cnt->vdev->params_array[indx].param_value);
-            }
-            if ( !strcmp(cnt->vdev->params_array[indx].param_name, "frequency")) {
-                frequency = atol(cnt->vdev->params_array[indx].param_value);
-            }
-        }
-
         bktr_method = METEOR_CAP_CONTINOUS;
 
         pthread_mutex_lock(&bktr_mutex);
@@ -866,11 +839,13 @@ int bktr_start(struct context *cnt)
         cnt->imgs.width = width;
         cnt->imgs.height = height;
 
-        cnt->vdev = mymalloc(sizeof(struct vdev_context));
-        memset(cnt->vdev, 0, sizeof(struct vdev_context));
-        cnt->vdev->usrctrl_array = NULL;
-        cnt->vdev->usrctrl_count = 0;
-        cnt->vdev->update_parms = TRUE;     /*Set trigger that we have updated user parameters */
+        cnt->vdev = mymalloc(sizeof(struct params_context));
+        memset(cnt->vdev, 0, sizeof(struct params_context));
+        cnt->vdev->params_array = NULL;
+        cnt->vdev->params_count = 0;
+        cnt->vdev->update_params = TRUE;     /*Set trigger to update parameters */
+
+        vid_parms_parse(cnt);
 
         /*
         * First we walk through the already discovered video devices to see
@@ -892,7 +867,7 @@ int bktr_start(struct context *cnt)
                 MOTION_LOG(NTC, TYPE_VIDEO, NO_ERRNO
                     ,_("Reusing [%s] inputs [%d,%d] Change capture"
                     " method METEOR_CAP_SINGLE")
-                    , dev->video_device, dev->input, conf->input);
+                    , dev->video_device, dev->input, cnt->param_input);
 
                 dev->bktr_method = METEOR_CAP_SINGLE;
 
@@ -923,7 +898,9 @@ int bktr_start(struct context *cnt)
         }
 
         /* Only open tuner if conf->tuner_device has set , freq and input is 1. */
-        if ((conf->tuner_device != NULL) && (frequency > 0) && (input == BKTR_IN_TV)) {
+        if ((conf->tuner_device != NULL) &&
+            (cnt->param_freq > 0) &&
+            (cnt->param_input == BKTR_IN_TV)) {
             bktr_fdtuner = open(conf->tuner_device, O_RDWR|O_CLOEXEC);
             if (bktr_fdtuner < 0) {
                 MOTION_LOG(CRT, TYPE_VIDEO, SHOW_ERRNO,_("open tuner device %s"),
@@ -942,10 +919,10 @@ int bktr_start(struct context *cnt)
         dev->bktr_tuner = conf->tuner_device;
         dev->fd_device = fd_device;
         dev->bktr_fdtuner = bktr_fdtuner;
-        dev->input = input;
+        dev->input = cnt->param_input;
         dev->height = height;
         dev->width = width;
-        dev->frequency = frequency;
+        dev->frequency = cnt->param_freq;
         dev->owner = -1;
         dev->bktr_method = bktr_method;
 
@@ -961,7 +938,7 @@ int bktr_start(struct context *cnt)
         dev->bktr_curbuffer = 0;
         dev->bktr_maxbuffer = 1;
 
-        if (!bktr_device_init(dev, width, height, input, norm, frequency)) {
+        if (!bktr_device_init(dev, width, height, cnt->param_input, cnt->param_norm, cnt->param_freq)) {
             close(dev->fd_device);
             pthread_mutexattr_destroy(&dev->attr);
             pthread_mutex_destroy(&dev->mutex);
@@ -1026,8 +1003,8 @@ int bktr_next(struct context *cnt,  struct image_data *img_data)
             dev->frames = conf->roundrobin_frames;
         }
 
-        bktr_set_input(cnt, dev, img_data->image_norm, width, height, conf->input, conf->norm,
-                    conf->roundrobin_skip, conf->frequency);
+        bktr_set_input(cnt, dev, img_data->image_norm, width, height, cnt->param_input,
+                       cnt->param_norm, conf->roundrobin_skip, cnt->param_freq);
 
         ret = bktr_capture(dev, img_data->image_norm, width, height);
 

--- a/src/video_common.c
+++ b/src/video_common.c
@@ -486,11 +486,13 @@ void vid_greytoyuv420p(unsigned char *map, unsigned char *cap_map, int width, in
 /* vid_parms_parse
  * Parse the video_params into an array.
 */
-int vid_parms_parse(struct context *cnt)
+void vid_parms_parse(struct context *cnt)
 {
 
+    int indx;
+
     if (cnt->vdev->update_params == FALSE) {
-        return 0;
+        return;
     }
 
     /* Put in the user specified parameters */
@@ -502,9 +504,21 @@ int vid_parms_parse(struct context *cnt)
     util_parms_add_default(cnt->vdev,"norm","0");
     util_parms_add_default(cnt->vdev,"frequency","0");
 
+    for (indx = 0; indx < cnt->vdev->params_count; indx++) {
+        if (!strcmp(cnt->vdev->params_array[indx].param_name, "input")) {
+            cnt->param_input = atoi(cnt->vdev->params_array[indx].param_value);
+        }
+        if (!strcmp(cnt->vdev->params_array[indx].param_name, "norm")) {
+            cnt->param_norm = atoi(cnt->vdev->params_array[indx].param_value);
+        }
+        if (!strcmp(cnt->vdev->params_array[indx].param_name, "frequency")) {
+            cnt->param_freq = atol(cnt->vdev->params_array[indx].param_value);
+        }
+    }
+
     cnt->vdev->update_params = FALSE;
 
-    return 0;
+    return;
 
 }
 

--- a/src/video_common.h
+++ b/src/video_common.h
@@ -56,7 +56,7 @@ void vid_close(struct context *cnt);
 void vid_mutex_destroy(void);
 void vid_mutex_init(void);
 
-int vid_parms_parse(struct context *cnt);
+void vid_parms_parse(struct context *cnt);
 
 void vid_yuv422to420p(unsigned char *map, unsigned char *cap_map, int width, int height);
 void vid_yuv422pto420p(unsigned char *map, unsigned char *cap_map, int width, int height);

--- a/src/video_v4l2.c
+++ b/src/video_v4l2.c
@@ -145,7 +145,7 @@ static void v4l2_vdev_free(struct context *cnt)
     }
 }
 
-static int v4l2_vdev_init(struct context *cnt)
+static void v4l2_vdev_init(struct context *cnt)
 {
 
     /* Create the v4l2 context within the main thread context  */
@@ -155,7 +155,7 @@ static int v4l2_vdev_init(struct context *cnt)
     cnt->vdev->params_count = 0;
     cnt->vdev->update_params = TRUE;     /*Set trigger that we have updated user parameters */
 
-    return 0;
+    return;
 
 }
 
@@ -528,25 +528,16 @@ static int v4l2_input_select(struct context *cnt, struct video_dev *curdev)
     /* Set the input number for the device if applicable */
     src_v4l2_t *vid_source = (src_v4l2_t *) curdev->v4l2_private;
     struct v4l2_input    input;
-    int indx, tmpinp;
 
-    tmpinp = -1;
-    for (indx = 0;indx < cnt->vdev->params_count; indx++) {
-        if ( !strcmp(cnt->vdev->params_array[indx].param_name,"input")) {
-            tmpinp = atoi(cnt->vdev->params_array[indx].param_value);
-            break;
-        }
-    }
-
-    if ((tmpinp == curdev->input) && (!curdev->starting)) {
+    if ((cnt->param_input == curdev->input) && (!curdev->starting)) {
         return 0;
     }
 
     memset(&input, 0, sizeof (struct v4l2_input));
-    if (tmpinp == -1 ) {
+    if (cnt->param_input == -1 ) {
         input.index = 0;
     } else {
-        input.index = tmpinp;
+        input.index = cnt->param_input;
     }
 
     if (xioctl(vid_source, VIDIOC_ENUMINPUT, &input) == -1) {
@@ -578,7 +569,7 @@ static int v4l2_input_select(struct context *cnt, struct video_dev *curdev)
         return -1;
     }
 
-    curdev->input        = tmpinp;
+    curdev->input        = cnt->param_input;
     curdev->device_type  = input.type;
     curdev->device_tuner = input.tuner;
 
@@ -592,17 +583,8 @@ static int v4l2_norm_select(struct context *cnt, struct video_dev *curdev)
     src_v4l2_t *vid_source = (src_v4l2_t *) curdev->v4l2_private;
     struct v4l2_standard standard;
     v4l2_std_id std_id;
-    int norm, indx;
 
-    norm = 0;
-    for (indx = 0;indx < cnt->vdev->params_count; indx++) {
-        if ( !strcmp(cnt->vdev->params_array[indx].param_name,"norm")) {
-            norm = atoi(cnt->vdev->params_array[indx].param_value);
-            break;
-        }
-    }
-
-    if ((norm == curdev->norm) && (!curdev->starting)) {
+    if ((cnt->param_norm == curdev->norm) && (!curdev->starting)) {
         return 0;
     }
 
@@ -611,7 +593,7 @@ static int v4l2_norm_select(struct context *cnt, struct video_dev *curdev)
             MOTION_LOG(NTC, TYPE_VIDEO, NO_ERRNO
                 ,_("Device does not support specifying PAL/NTSC norm"));
         }
-        norm = std_id = 0;    // V4L2_STD_UNKNOWN = 0
+        cnt->param_norm = std_id = 0;    // V4L2_STD_UNKNOWN = 0
     }
 
     if (std_id) {
@@ -626,7 +608,7 @@ static int v4l2_norm_select(struct context *cnt, struct video_dev *curdev)
             standard.index++;
         }
 
-        switch (norm) {
+        switch (cnt->param_norm) {
         case 1:
             std_id = V4L2_STD_NTSC;
             break;
@@ -654,7 +636,7 @@ static int v4l2_norm_select(struct context *cnt, struct video_dev *curdev)
         }
     }
 
-    curdev->norm = norm;
+    curdev->norm = cnt->param_norm;
 
     return 0;
 }
@@ -666,18 +648,8 @@ static int v4l2_frequency_select(struct context *cnt, struct video_dev *curdev)
     src_v4l2_t *vid_source = (src_v4l2_t *) curdev->v4l2_private;
     struct v4l2_tuner     tuner;
     struct v4l2_frequency freq;
-    int indx;
-    long usrfreq;
 
-    usrfreq = 0;
-    for (indx = 0;indx < cnt->vdev->params_count; indx++) {
-        if ( !strcmp(cnt->vdev->params_array[indx].param_name,"frequency")) {
-            usrfreq = atol(cnt->vdev->params_array[indx].param_value);
-            break;
-        }
-    }
-
-    if ((usrfreq == curdev->frequency) && (!curdev->starting)) {
+    if ((cnt->param_freq == curdev->frequency) && (!curdev->starting)) {
         return 0;
     }
 
@@ -701,7 +673,7 @@ static int v4l2_frequency_select(struct context *cnt, struct video_dev *curdev)
         memset(&freq, 0, sizeof(struct v4l2_frequency));
         freq.tuner = curdev->device_tuner;
         freq.type = V4L2_TUNER_ANALOG_TV;
-        freq.frequency = (usrfreq / 1000) * 16;
+        freq.frequency = (cnt->param_freq / 1000) * 16;
 
         if (xioctl(vid_source, VIDIOC_S_FREQUENCY, &freq) == -1) {
             MOTION_LOG(ERR, TYPE_VIDEO, SHOW_ERRNO
@@ -714,7 +686,7 @@ static int v4l2_frequency_select(struct context *cnt, struct video_dev *curdev)
         }
     }
 
-    curdev->frequency = usrfreq;
+    curdev->frequency = cnt->param_freq;
 
     return 0;
 }
@@ -1178,7 +1150,6 @@ static int v4l2_device_init(struct context *cnt, struct video_dev *curdev)
 {
 
     src_v4l2_t *vid_source;
-    int indx;
 
     /* Allocate memory for the state structure. */
     if (!(vid_source = calloc(sizeof(src_v4l2_t), 1))) {
@@ -1191,19 +1162,9 @@ static int v4l2_device_init(struct context *cnt, struct video_dev *curdev)
     pthread_mutex_init(&curdev->mutex, &curdev->attr);
 
     curdev->usage_count = 1;
-
-    for (indx = 0; indx < cnt->vdev->params_count; indx++) {
-        if ( !strcmp(cnt->vdev->params_array[indx].param_name, "input")) {
-            curdev->input = atoi(cnt->vdev->params_array[indx].param_value);
-        }
-        if ( !strcmp(cnt->vdev->params_array[indx].param_name, "norm")) {
-            curdev->norm = atoi(cnt->vdev->params_array[indx].param_value);
-        }
-        if ( !strcmp(cnt->vdev->params_array[indx].param_name, "frequency")) {
-            curdev->frequency = atol(cnt->vdev->params_array[indx].param_value);
-        }
-    }
-
+    curdev->input = cnt->param_input;
+    curdev->norm = cnt->param_norm;
+    curdev->frequency = cnt->param_freq;
     curdev->height = cnt->conf.height;
     curdev->width = cnt->conf.width;
 
@@ -1233,20 +1194,14 @@ static void v4l2_device_select(struct context *cnt, struct video_dev *curdev, un
         return;
     }
 
-    newvals = FALSE;
-    for (indx = 0; indx < cnt->vdev->params_count; indx++) {
-        if (!strcmp(cnt->vdev->params_array[indx].param_name, "input") &&
-            atoi(cnt->vdev->params_array[indx].param_value) != curdev->input) {
-            newvals = TRUE;
-        }
-        if (!strcmp(cnt->vdev->params_array[indx].param_name, "norm") &&
-            atoi(cnt->vdev->params_array[indx].param_value) != curdev->norm) {
-            newvals = TRUE;
-        }
-        if (!strcmp(cnt->vdev->params_array[indx].param_name, "frequency") &&
-            atol(cnt->vdev->params_array[indx].param_value) != curdev->frequency) {
-            newvals = TRUE;
-        }
+    vid_parms_parse(cnt);
+
+    if ((cnt->param_input != curdev->input) ||
+        (cnt->param_norm != curdev->norm) ||
+        (cnt->param_freq != curdev->frequency)) {
+        newvals = TRUE;
+    } else {
+        newvals = FALSE;
     }
 
     if (newvals) {
@@ -1256,9 +1211,6 @@ static void v4l2_device_select(struct context *cnt, struct video_dev *curdev, un
         }
         if (retcd == 0) {
             retcd = v4l2_frequency_select(cnt, curdev);
-        }
-        if (retcd == 0) {
-            retcd = vid_parms_parse(cnt);
         }
         if (retcd == 0) {
             retcd = v4l2_parms_set(cnt, curdev);
@@ -1286,10 +1238,7 @@ static void v4l2_device_select(struct context *cnt, struct video_dev *curdev, un
 
     } else {
         /* No round robin - we only adjust picture controls */
-        retcd = vid_parms_parse(cnt);
-        if (retcd == 0) {
-            retcd = v4l2_parms_set(cnt, curdev);
-        }
+        retcd = v4l2_parms_set(cnt, curdev);
         if (retcd == 0) {
             retcd = v4l2_autobright(cnt, curdev, cnt->conf.auto_brightness);
         }
@@ -1306,19 +1255,12 @@ static void v4l2_device_select(struct context *cnt, struct video_dev *curdev, un
 static int v4l2_device_open(struct context *cnt, struct video_dev *curdev)
 {
 
-    int fd_device, indx, usrinp;
+    int fd_device;
     /* Open the video device */
-
-    usrinp = -1;
-    for (indx = 0; indx < cnt->vdev->params_count; indx++) {
-        if ( !strcmp(cnt->vdev->params_array[indx].param_name, "input")) {
-            usrinp = atoi(cnt->vdev->params_array[indx].param_value);
-        }
-    }
 
     MOTION_LOG(NTC, TYPE_VIDEO, NO_ERRNO
         ,_("Using videodevice %s and input %d")
-        ,cnt->conf.video_device, usrinp);
+        ,cnt->conf.video_device, cnt->param_input);
 
     /* Give the watchdog more time for this open function */
     cnt->watchdog = (WATCHDOG_TMO * 2);
@@ -1518,13 +1460,12 @@ int v4l2_start(struct context *cnt)
         curdev = video_devices;
         while (curdev) {
             if (!strcmp(cnt->conf.video_device, curdev->video_device)) {
-                retcd = v4l2_vdev_init(cnt);
-                if (retcd == 0) {
-                    retcd = vid_parms_parse(cnt);
-                }
-                if (retcd == 0) {
-                    retcd = v4l2_imgs_set(cnt, curdev);
-                }
+                v4l2_vdev_init(cnt);
+
+                vid_parms_parse(cnt);
+
+                retcd = v4l2_imgs_set(cnt, curdev);
+
                 if (retcd == 0) {
                     curdev->usage_count++;
                     retcd = curdev->fd_device;
@@ -1539,13 +1480,11 @@ int v4l2_start(struct context *cnt)
 
         curdev->starting = TRUE;
 
-        retcd = v4l2_vdev_init(cnt);
-        if (retcd == 0) {
-            retcd = vid_parms_parse(cnt);
-        }
-        if (retcd == 0) {
-            retcd = v4l2_device_init(cnt, curdev);
-        }
+        v4l2_vdev_init(cnt);
+
+        vid_parms_parse(cnt);
+
+        retcd = v4l2_device_init(cnt, curdev);
         if (retcd == 0) {
             retcd = v4l2_device_open(cnt, curdev);
         }

--- a/src/webu.c
+++ b/src/webu.c
@@ -700,7 +700,7 @@ static int webu_process_config_set(struct webui_ctx *webui)
                 , webui->uri_parm1, webui->uri_value1);
 
             /*If we are updating vid parms, set the flag to update the device.*/
-            if (!strcmp(config_params[indx].param_name, "vid_control_params") &&
+            if (!strcmp(config_params[indx].param_name, "video_params") &&
                 (webui->cntlst[webui->thread_nbr]->vdev != NULL)) {
                 webui->cntlst[webui->thread_nbr]->vdev->update_params = TRUE;
             }

